client/src/main/java/io/pravega/client/tables/KeyValueTable.java

```diff
@@ -34,58 +32,84 @@
- * Key-Value Table are associated with a single Key Family; in this case the entire Key-Value Table load will be placed
- * on a single backing Table Segment instead of spreading it across many Table Segments, leading to eventual performance
- * degradation.
+ * <li> The Primary Key is mandatory and is used to group all {@link TableKey}s with the same Primary Key together in the
+ * same Table Partition; this allows multiple keys/entries sharing the same Primary Key to be updated/removed atomically.
+ * The Primary Keys must all have the same length throughout the Key-Value Table. This needs to be configured when the
+ * Key-Value Table is created (See {@link KeyValueTableConfiguration#getPrimaryKeyLength()} and must be greater than 0.
+ *
+ * <li> The Secondary Key is optional. The Secondary Keys must all have the same length throughout the Key-Value Table.
+ * This needs to be configured when the Key-Value Table is created (See {@link KeyValueTableConfiguration#getSecondaryKeyLength()}
+ * and must be at least 0. For Key-Value Tables with no Secondary Keys, set this value to 0.
+ *
+ * <li> The Primary Keys must all have the same length throughout the Key-Value Table. This needs to be configured when
+ * the Key-Value Table is created (See {@link KeyValueTableConfiguration#getPrimaryKeyLength()}.
+ *
+ * <li> Two {@link TableKey} instances are considered equal if they have the same Primary Key and same Secondary Key.
+ * The same Secondary Key may be "associated" with different Primary Keys, in which case the resulting {@link TableKey}s
+ * will be distinct. For example, consider Primary Keys PK1 != PK2 and Secondary Key SK1; key PK1.SK1 is different from
+ * PK2.SK1 and both are different from PK1 and PK2 (with no secondary key).
+ *
+ * <li> Multiple {@link TableKey}/{@link TableEntry} instances with the same Primary Key can be updated or removed
+ * atomically (either all changes will be applied or none will).
+ *
+ * <li> {@link TableKey}s that do not share the same Primary Key may be hashed to different Key-Value Table Partitions
+ * and cannot be used for multi-key/entry atomic updates or removals.
+ *
+ * <li> {@link TableKey}s sharing the same Primary Key are grouped into the same Table Segment; as such, the choice of
+ * Primary Keys can have performance implications. An ideally balanced Key-Value Table is one where there the load on each
+ * Table Segment is approximately the same; this can be achieved by not having any Secondary Keys at all or where each
+ * Primary Key has approximately the same number of Secondary Keys. To enable a uniform distribution of Keys over the
+ * Key-Value Table, consider not using Secondary Keys at all. If this situation cannot be avoided (i.e., multi-entry
+ * atomic updates are required), then it is recommended to use a sufficiently large number of Primary Keys compared to
+ * the number of Partitions configured on the Key-Value Table (the higher the ratio of Primary Keys to Partitions, the
```

Comment:
Both are valid points. @fpj or @tkaitchuck  do you have any suggestions on how I could improve the wording here, if at all?

Reply:
One aspect that stands out to me is that the application is not necessarily free to choose keys as they are part of the applications semantics and the key distribution is driven by the workload. What we can tell the application developer is how to configure a KVT depending on the type of use case, but the only knob I actually see with respect to the distribution of keys is the number of partitions. Choosing a larger number of partitions won't necessarily help with distributing the load more evenly. In fact, the degree of parallelism is upper bounded by the number of segment stores for reads and number of segment containers for writes, but we make no guarantee about how the load is distributed.

The above refers to accesses to individual pairs. The other point I touched upon is how to efficiently process a whole table and in particular what we offer that helps the application split the workload if needed.

Reply:
I'm not necessarily suggesting to replace it verbatim, but I thought we could say something like:

```
<li> {@link TableKey}s sharing the same Primary Key are grouped into the same Table Segment; as such, the choice of
 * table key can have performance implications. As Primary Keys are hashed to Table Segments, a large number of Primary
 * Keys (larger than the number of Table Segments) and a uniform distribution of operations across keys leads to an even
 * distribution of load across Table Segments. Many realistic workloads are skewed, however, and can lead to an uneven
 * load distribution across Table Segments.
 *
 * The use of Secondary Keys has also implications to load balance. All keys containing a given Primary Key map to the
 * same Table Segment independent of Secondary Key. In the presence of Secondary Keys, the total load for a given
 * Primary Key depends on the load aggregate across all of its Secondary Keys. When the load distribution is highly
 * skewed, one option to consider it eliminating Secondary Keys and only using Primary Keys. The application must be
 * able to encode the key into a single one rather than split into two parts. To illustrate, if we have two Table Keys
 * "A.B" and "A.C", "." representing the separation between Primary and Secondary Keys, then they map to the same Table
 * Segment because they have the same Primary Key "A". If instead, we use "AB" and "AC" as keys, eliminating the Secondary
 * Keys but retaining the necessary information in the Primary Key, then we should be able to map those keys to different
 * Table Segments depending on the hashing.
 
 ```

Reply:
Fixed.